lib/IRGen/GenType.cpp

```diff
@@ -1797,6 +1797,57 @@ static bool isIRTypeDependent(IRGenModule &IGM, NominalTypeDecl *decl) {
   }
 }
 
+bool irgen::mightContainMetadata(const IRGenModule &IGM, const CanType type) {
+  IRGenModule &IGMCast = const_cast<IRGenModule &>(IGM);
+  auto runType = irgen::getRuntimeReifiedType(IGMCast, type);
+  if (isa<SILFunctionType>(runType)) {
+    return false;
+  }
+  if (!IsIRTypeDependent(IGMCast).visit(runType)) {
+    return false;
+  }
+  switch (runType->getKind()) {
+#define SUGARED_TYPE(ID, SUPER) case TypeKind::ID:
+#define UNCHECKED_TYPE(ID, SUPER) case TypeKind::ID:
+#define TYPE(ID, SUPER)
+#include "swift/AST/TypeNodes.def"
+  case TypeKind::Error:
+    llvm_unreachable("kind is invalid for a canonical type");
+
+#define ARTIFICIAL_TYPE(ID, SUPER) case TypeKind::ID:
+#define TYPE(ID, SUPER)
+#include "swift/AST/TypeNodes.def"
+  case TypeKind::LValue:
+  case TypeKind::InOut:
+  case TypeKind::DynamicSelf:
+    return false;
+
+  case TypeKind::Tuple: {
+    auto genTuple = cast<TupleType>(runType);
+    for (auto elt : genTuple->getElements()) {
+      if (elt.getType()->getKind() == TypeKind::SILFunction) {
+        return false;
+      }
+    }
+    return true;
+  }
+  case TypeKind::BoundGenericEnum: {
+    auto genEnum = cast<BoundGenericEnumType>(runType);
+    for (auto arg : genEnum->getGenericArgs()) {
+      if (!arg->isCanonical()) {
+        continue;
+      }
+      if (mightContainMetadata(IGM, arg->getCanonicalType())) {
+        return true;
+      }
+    }
+    return false;
+  }
+  default:
+    return true;
```

Comment:
Exactly for what types does this return true? It's really unclear

Reply:
I will make it clearer / add the types I expect explicitly instead of catch-all default.

Reply:
In general, if I have code like

```
func foo<T, U, V>() {
  let value: (T -> U, Array<V>, Dictionary<Optional<T>, U>) = ...

  let copy = value
}
```

You want to outline the copy_addr call for the 'let copy', right? It seems like you should just walk the type `(T -> U, Array<V>, Dictionary<Optional<T>, U>)` recursively to collect all type parameters that appear in it, and create an outlined copy_addr that takes metadata for T, U, V as arguments. What am I missing here?